### ISSUE TITLE
One-time use variable: new setting to opt-in/out for return statements

### ISSUE THREAD

**Issue by kalessil** [kalessil/phpinspectionsea#1410]

**Comment by NamelessCoder** [comment]
Generally speaking, single-use variables aren't bad - they're good. It just depends on what you're doing - reading or debugging.

When you're debugging, having as many single-use variables as possible is a good thing and desirable, precisely because you can inspect a variable that is then fed to another function. This aspect becomes even more important if methods you call on object instances will mutate the object.

There is also the concern that (recursively) removing single-use variables can lead to function call nesting that instead of reducing mental complexity, drastically increases it - while also sacrificing debug ability.

I would be concerned that adding an opt-out for a specific context opens the door so you have to add opt-outs for many other contexts or depend on other inspections (max nested function calls etc.).

I honestly think that this particular inspection shouldn't be included at all.

**Comment by orklah** [comment]
@NamelessCoder I was going to respond that I kinda like this inspection but when I look back, there are many cases I don't apply it (which means I don't enable it in the inspection profile my team use):
- When I have to name a variable to explain why I do something, for example:
```php
$is_color_white = ($color === "#FFFFFF");
if($is_color_white)
```
I prefer this instead of getting rid of the variable and adding a comment
- When I have to add a docblock for the variable to help tools:
```php
function a():SpecificObject{
/** @var SpecificObject $a **/
$a = factory::getObject('SpecificObject');
return $a;
}
```
- If the variable reduce the complexity or length of a line:
```php
$list = $a->get_b()->get_c()->get_d()->get_e()->get_f()->get_list_of_things();
foreach($list as $elem){

}
//instead of 
foreach($a->get_b()->get_c()->get_d()->get_e()->get_f()->get_list_of_things() as $elem){

}
```

But I think there is one case that I would still like to see everytime:
```php
$a = $b;
if($a) // I still want the inspection to tell me that $a is useless here
```

**Comment by NamelessCoder** [comment]
@orklah I think you may be overlooking something:

> But I think there is one case that I would still like to see everytime:
> $a = $b;
> if($a) // I still want the inspection to tell me that $a is useless here

Unless, as you already mentioned, the variable that was assigned actually explains something about the condition - such as `$is_color_white`. So you probably don't want to see this inspection *every* time, which also hints to the concern I aired that you'd need so many context dependent behaviors that the inspection ends up making more trouble than it solves.

I agree with all your other points.

**Comment by orklah** [comment]
Yeah, my $a = $b example was to reflect a code that literally put a variable into another, no other operation.

If the need is just to clarify the content of the variable, it would be best to rename $b and keep using it instead of duplicating variables

**Comment by kalessil** [comment]
Introduced new settings.

PS: @orklah, I like the idea with if-statement, would you please create a separate ticket for it?
